### PR TITLE
Enhance type and argument manipulation in elementwise and reduction kernels

### PR DESCRIPTION
**Pull request by niboshi**
Separated from #2920.

Introduced `_ArgInfo` to hold argument metadata, and `_TypeMap` to abstract C type mapping.

### CODE REVIEW

**Comment by emcastillo** [comment]
There is no dependence with other prs, right?

**Comment by niboshi** [comment]
No, this PR can be reviewed independently. 👍

File: cupy/core/_kernel.pxd [hunk 1]
```diff
@@ -51,9 +104,9 @@ cdef class _Ops:
 cpdef create_ufunc(name, ops, routine=*, preamble=*, doc=*,
                    default_casting=*, loop_prep=*, out_ops=*)
 
-cpdef tuple _get_args_info(list args)
+cpdef tuple _get_arginfos(list args)
 
-cpdef str _get_kernel_params(tuple params, tuple args_info)
+cpdef str _get_kernel_params(tuple params, tuple arginfos)
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/core/_kernel.pxd]: I thought `arginfo` and `arginfos` are more consistent and easily guessed.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
It looks the error is unrelated?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
seems like pypi is down.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**